### PR TITLE
Apply ITEM_SPAWNRATE to monster drops

### LINKED ISSUE #28919: Ability to tune down/up monster drops
**Is your feature request related to a problem? Please describe.**  
Currently, you can only modify item spawn rates for items spawning in buildings, and there is no similar setting for monster drops. When playing with lowered item spawn rate settings, amount of loot dropped by monsters is unbalanced.

**Describe the solution you'd like**  
Additional setting for tuning monster drops up or down, similar to "Item spawn scaling factor".

**Describe alternatives you've considered**  
Having "Item spawn scaling factor" influence both building and monster drops. That's one possible solution, but gives less freedom for players who would like to have separate ratios for them.

### PR DESCRIPTION
**Pull request by Davi-DeGanne**
#### Summary
```SUMMARY: Balance "Apply item spawn scaling factor to monster drops"```

#### Purpose of change
Previously, setting a low 'item spawn scaling factor' did not affect drops from monsters, making a disproportionate amount of a player's loot come from monster drops.

Fixes #28919 

#### Describe the solution
Monster drops are now reduced as well, using the same setting as for standard item spawns, the 'item spawn scaling factor' as it is displayed in the world gen menu (ITEM_SPAWNRATE in the code). 

Example to illustrate how it works: If 'item spawn scaling factor' is set to 0.30, and a zombie dies, then each item it would _normally_ drop (i.e. lighter, cell phone, shirt, pants, shoes) individually has only a 30% chance of actually spawning. (NB: Corpses, broken robots, and other things spawned via the `revert_to_itype` JSON property are unaffected)

#### Describe alternatives you've considered
All the ideas mentioned in #28919, but this solution seemed like the compromise the most people agreed on.

We could also add a JSON flag where certain items are excluded, and *always* spawn, even on low item spawn settings, but:

- `revert_to_itype` already fills this role (though it isn't particularly robust).
- Even if people do want this, it seemed to me like it ought to be a separate PR.

#### Additional context
~~This might need more testing; I did do so myself, and caught a couple edge case bugs, but given how rusty I am with C++, I wouldn't be surprised if there are more.~~
Rearranged stuff so I cull the items before they spawn rather than after (don't ask me why I didn't just do it that way in the first place). After some more time testing the new code, I am now quite confident it is bug-free.

### CODE REVIEW

**Comment by John-Candlebury** [comment]
If you play with say 4x item spawns will a zombie drop say 4 pairs of pants? 

I think  that, ideally, that shouldn’t happen

**Comment by DaviBones** [comment]
> 
> 
> If you play with say 4x item spawns will a zombie drop say 4 pairs of pants?
> 
> I think that, ideally, that shouldn’t happen

Nope, my code is completely bypassed if `ITEM_SPAWNRATE >= 1`, so same behavior as before.

File: src/monster.cpp [hunk 1]
```diff
@@ -2250,8 +2250,27 @@ void monster::drop_items_on_death()
...
+    if( spawn_rate < 1 ) {
+        // Temporary vector, to remember which items will be dropped
+        std::vector<item> remaining;
+        for( item it : items ) {
+            if( rng_float( 0, 1 ) < spawn_rate ) {
+                remaining.push_back( it );
+            }
+        }
+        // If there aren't any items left, there's nothing left to do
+        if( remaining.size() == 0 ) {
```
Comment (BevapDin) [inline src/monster.cpp]:
```suggestion
        if( remaining.empty() ) {
```

File: src/monster.cpp [hunk 2]
```diff
@@ -2250,8 +2250,27 @@ void monster::drop_items_on_death()
...
-                         calendar::start_of_cataclysm );
+
+    std::vector<item> items = item_group::items_from( type->death_drops, calendar::start_of_cataclysm );
+
+    // This block removes some items, according to item spawn scaling factor
+    const float spawn_rate = get_option<float>( "ITEM_SPAWNRATE" );
+    if( spawn_rate < 1 ) {
+        // Temporary vector, to remember which items will be dropped
+        std::vector<item> remaining;
+        for( item it : items ) {
```
Comment (BevapDin) [inline src/monster.cpp]:
```suggestion
        for( const item &it : items ) {
```

Comment (DaviBones) [reply src/monster.cpp]: I understand why `const` is good practice here, but would you mind explaining why we want to use an address here? (sorry, bit of a c++ noob)

Comment (kevingranade) [reply src/monster.cpp]: If it captures a reference, it does not make a copy. The item class is large enough that it's imperative to avoid making copies unecesssrally.

Comment (DaviBones) [reply src/monster.cpp]: Oh, holy crap! I didn't know it made a copy! Yeah, I definitely see why that's better now. Thanks.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/tips-tricks-and-newb-questions-reborn/16870/1355
